### PR TITLE
Metrics - modules decide what metrics to report

### PR DESCRIPTION
**Pull request by janciesla8818**
This PR is implementing changes to the system metrics functionality. New behaviour `mongoose_module_metrics` is implemented and there are changes to the `mongoose_system_metrics_collector`.

Updates to `mongoose_system_metrics_collector`:
* only modules that are implementing `mongoose_module_metrics` behaviour are reported
* counting and reporting number of custom modules

`mongoose_module_metrics` behaviour:
* modules that implement this behaviour are reported
* some modules implement additional callback that allows to report additional options that the module is configured with (e.g. backend)

### CODE REVIEW

**Comment by mongoose-im** [comment]
[7727.1](https://travis-ci.org/esl/MongooseIM/jobs/636368896) / Erlang 22.0 / small_tests / 7afe3d05d8de43b32f1153aaabb3ea12a6d8eba0
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2591/7727/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2591/7727/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-01-13_14.01.24/lib.mongooseim.logs/run.2020-01-13_14.01.24/suite.log.html)

**Comment by mongoose-im** [comment]
[7730.1](https://travis-ci.org/esl/MongooseIM/jobs/636407339) / Erlang 22.0 / small_tests / 3dcc530c2bd5e3fc9b9319de266b5b4d82fc5219
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2591/7730/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2591/7730/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-01-13_16.11.43/lib.mongooseim.logs/run.2020-01-13_16.11.44/suite.log.html)

**Comment by mongoose-im** [comment]
[7732.1](https://travis-ci.org/esl/MongooseIM/jobs/636459802) / Erlang 22.0 / small_tests / b8bf40ca28c66388df43247a5d26560b3c9e000c
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2591/7732/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2591/7732/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-01-13_17.21.36/lib.mongooseim.logs/run.2020-01-13_17.21.36/suite.log.html)

File: src/system_metrics/mongoose_system_metrics_collector.erl [hunk 1]
```diff
@@ -33,10 +34,55 @@ get_hosts_count() ->
...
+    Hosts = ejabberd_config:get_global_option(hosts),
+    AllModules = lists:flatten(
+                    lists:map(fun gen_mod:loaded_modules/1, Hosts)),
+    GenMods = filter_behaviour_implementations(AllModules, gen_mod),
+    GenModsSet = sets:from_list(GenMods),
+    MetricsModule = filter_behaviour_implementations(AllModules,
+                                                     mongoose_module_metrics),
+    MetricsModuleSet = sets:from_list(MetricsModule),
+    CountCustomMods= sets:size(sets:subtract(GenModsSet, MetricsModuleSet)),
+    #{report_name => custom_modules_names, key => sets:to_list(sets:subtract(GenModsSet, MetricsModuleSet)), value => CountCustomMods}.
```
Comment (NelsonVides) [inline src/system_metrics/mongoose_system_metrics_collector.erl]: 😱

File: src/system_metrics/mongoose_module_metrics.erl [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+-optional_callbacks([config_metrics/1]).
+
+opts_for_module(Host, Module, OptsToReport) ->
+    try
+        Opts = gen_mod:opts_for_module(Host, Module),
+        lists:map(
+            fun({OptToReport, DefaultValue}) ->
+                    Value = proplists:get_value(OptToReport, Opts, DefaultValue),
+                    {OptToReport, Value}
+                end,OptsToReport)
```
Comment (NelsonVides) [inline src/system_metrics/mongoose_module_metrics.erl]: Broken indentation, should be like 
```erlang
        lists:map(
            fun({OptToReport, DefaultValue}) ->
                    Value = proplists:get_value(OptToReport, Opts, DefaultValue),
                    {OptToReport, Value}
            end, OptsToReport)
```

File: src/system_metrics/mongoose_system_metrics_collector.erl (hunk 1 above)
Comment (janciesla8818) [reply src/system_metrics/mongoose_system_metrics_collector.erl]: work in progress, the line will be different after testing is copleted

**Comment by mongoose-im** [comment]
[7736.1](https://travis-ci.org/esl/MongooseIM/jobs/636912211) / Erlang 22.0 / small_tests / 0eff03162a5115b10d74b04a05d60f1431230f18
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2591/7736/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2591/7736/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-01-14_14.25.09/lib.mongooseim.logs/run.2020-01-14_14.25.09/suite.log.html)

**Comment by mongoose-im** [comment]
[7737.1](https://travis-ci.org/esl/MongooseIM/jobs/636939342) / Erlang 22.0 / small_tests / f9ba979710334b75051b4774c7ac6456d9a4a2a0
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2591/7737/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2591/7737/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-01-14_15.35.24/lib.mongooseim.logs/run.2020-01-14_15.35.24/suite.log.html)

**Comment by mongoose-im** [comment]
[7745.1](https://travis-ci.org/esl/MongooseIM/jobs/637002827) / Erlang 22.0 / small_tests / 08122753f60a41fa06895ae6b17b5e575a84b70b
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2591/7745/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2591/7745/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-01-14_17.54.47/lib.mongooseim.logs/run.2020-01-14_17.54.47/suite.log.html)

**Pull request merged**